bus-mapping/src/circuit_input_builder.rs

```diff
@@ -244,3 +317,1048 @@ impl<'a> CircuitInputBuilder {
+        }
+    })
+}
+
+impl<'a> CircuitInputStateRef<'a> {
+    fn get_step_err(
+        &self,
+        step: &GethExecStep,
+        next_step: Option<&GethExecStep>,
+    ) -> Option<ExecError> {
+        if let Some(error) = get_step_reported_error(step) {
+            return Some(error);
+        }
+        assert!(step.error.is_none());
+
+        let next_depth = next_step.map(|s| s.depth).unwrap_or(0);
+        let next_result = next_step
+            .map(|s| s.stack.last().unwrap_or_else(|_| Word::zero()))
+            .unwrap_or_else(Word::zero);
+
+        // Return from a call with a failure, without calling RETURN
+        if step.op != OpcodeId::RETURN
+            && step.depth != next_depth
+            && next_result == Word::zero()
+        {
+            return Some(match step.op {
+                OpcodeId::REVERT => ExecError::ExecutionReverted,
+                OpcodeId::JUMP | OpcodeId::JUMPI => ExecError::InvalidJump,
+                OpcodeId::RETURNDATACOPY => ExecError::ReturnDataOutOfBounds,
+                _ => {
+                    panic!("Cannot figure out call failure error in {:?}", step)
+                }
+            });
+        }
+
+        // Return from a {CREATE, CREATE2} with a failure, via RETURN
+        // TODO: if not self.tx_ctx.is_root and self.tx_ctx.is_create {
+        if step.op == OpcodeId::RETURN && next_result == Word::zero() {
+            let offset = step.stack.nth_last(0).unwrap();
+            let length = step.stack.nth_last(1).unwrap();
+            if length > Word::from(0x6000) {
+                return Some(ExecError::MaxCodeSizeExceeded);
+            } else if length > Word::zero()
+                && !step.memory.0.is_empty()
+                && step.memory.0.get(offset.low_u64() as usize) == Some(&0xef)
+            {
+                return Some(ExecError::InvalidCode);
+            } else if Word::from(200) * length > Word::from(step.gas.0) {
+                return Some(ExecError::CodeStoreOutOfGas);
+            } else {
+                panic!("Cannot figure out RETURN error in {:?}", step);
+            }
+        }
+        // }
+
+        // The *CALL* code was not executed
+        let next_pc = next_step.map(|s| s.pc.0).unwrap_or(1);
+        if [
+            OpcodeId::CALL,
+            OpcodeId::CALLCODE,
+            OpcodeId::DELEGATECALL,
+            OpcodeId::STATICCALL,
+            OpcodeId::CREATE,
+            OpcodeId::CREATE2,
+        ]
+        .contains(&step.op)
+            && next_result == Word::zero()
+            && next_pc != 0
+        {
+            if step.depth == 1025 {
+                return Some(ExecError::Depth);
+            }
+            // TODO: address_collision
+            if step.op == OpcodeId::CREATE2 {
+                // NOTE: We don't expect Address Collision in CREATE because in
+                // that case, the Address is always unique (the
+                // probability of collision is negligible).
```

Comment:
I believe `CREATE` could still create same address especially a contract selfdestructed and then recreated using `CREATE2`

Reply:
If I undertand it correctly, you mean:
1. Create contract A with CREATE2
2. Contract A does CREATE with nonce N and ARGS -> new Address B
3. Contract A selfdestructs
4. Repeat steps 1-3: In step 2, Address B already exists

Well that makes sense :)

Reply:
Exactly! tho it should be a extremely rare case